Add zip option for packaging

t0mdavid-m opened the pull request:
### **User description**
## Description

This PR allows the creation of the installer as `.zip` file during the creation. This is especially relevant for portable applications (e.g. the windows version of the streamlit webapps), where all dependencies should be included. In order to build create a zip folder instead of a binary, the environment variable `PACK_ZIP` has to be set.

This PR is based on a discssion in #7528.

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).


___

### **PR Type**
Enhancement


___

### **Description**
- Removed hardcoded `CPACK_GENERATOR` settings for various package types (DEB, DragNDrop, productbuild, NSIS, RPM).
- Added logic in `CMakeLists.txt` to dynamically set `CPACK_GENERATOR` based on `PACKAGE_TYPE` or `PACK_ZIP` environment variable.
- Updated contrib submodule reference.



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>package_deb.cmake</strong><dd><code>Remove hardcoded DEB package generator setting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cmake/package_deb.cmake

- Removed hardcoded `CPACK_GENERATOR` setting for DEB packages.



</details>
    

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7529/files#diff-e1ab9f54a264ed2639a56781454247b865566873543fb034d3a1f20b9c2d3650">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>package_dragndrop_dmg.cmake</strong><dd><code>Remove hardcoded DragNDrop package generator setting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cmake/package_dragndrop_dmg.cmake

<li>Removed hardcoded <code>CPACK_GENERATOR</code> setting for DragNDrop packages.<br>


</details>
    

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7529/files#diff-5963b0dad0ddca4582d3cb94c61d678f15ac1cf9652a26d3f0cab26e8c7a8a92">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>package_mac_productbuild.cmake</strong><dd><code>Remove hardcoded productbuild package generator setting</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
cmake/package_mac_productbuild.cmake

<li>Removed hardcoded <code>CPACK_GENERATOR</code> setting for productbuild packages.<br>


</details>
    

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7529/files#diff-097ab080848fb655d79615761e081730143dfb70798cd79cc85c92585f87feec">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>package_nsis.cmake</strong><dd><code>Remove hardcoded NSIS package generator setting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cmake/package_nsis.cmake

- Removed hardcoded `CPACK_GENERATOR` setting for NSIS packages.



</details>
    

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7529/files#diff-cff983f198f56112242929418519d3f585d6c7787543b952e0cea257700bbcb0">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>package_rpm.cmake</strong><dd><code>Remove hardcoded RPM package generator setting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cmake/package_rpm.cmake

- Removed hardcoded `CPACK_GENERATOR` setting for RPM packages.



</details>
    

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7529/files#diff-e0f033c049997f539d53a3a3c59b5fbf0076f35eaebe574ee026276e5f4e63e0">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CMakeLists.txt</strong><dd><code>Add logic for dynamic CPack generator selection</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
CMakeLists.txt

<li>Added logic to set <code>CPACK_GENERATOR</code> based on <code>PACKAGE_TYPE</code> or <code>PACK_ZIP</code> <br>environment variable.<br> <li> Included conditional packaging routines based on <code>PACKAGE_TYPE</code>.<br>


</details>
    

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7529/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a">+21/-3</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Dependencies</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>contrib</strong><dd><code>Update contrib submodule reference</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
contrib

- Updated submodule commit reference.



</details>
    

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7529/files#diff-fc1e87ada6cfa1335e3062451e7ac9d13f17482c4660989eeee7d5ec75a620d6">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

t0mdavid-m committed:
support packaging as zip

jpfeuffer commented:
Makes sense! I didn't check the changes, yet, but we could also enable it with cmake -DPACKAGE_TYPE="zip".

This is our usual (wrapped) way of choosing the packaging type for cpack.

t0mdavid-m commented:
> Makes sense! I didn't check the changes, yet, but we could also enable it with cmake -DPACKAGE_TYPE="zip".
> 
> This is our usual (wrapped) way of choosing the packaging type for cpack.

I also thought about that. However, then we would need another variable to determine which cmake scripts to include. So unfortunately I think we need both.

jpfeuffer commented:
Do I understand correctly that you want to specify zip+nsis essentially? And you want to do it because the package_nsis creates some components that you want in the zip on windows? 

Maybe I would double check the package_xyz.cmake files then and pull out things that actually are dependent on the OS instead of the actual package type and put it into package_general.cmake.

I think I gradually tried to do that. Note that there is some code in other cmake sections that depends on packaging (e.g. Mac app bundle creation, which is a bit hindering during development). This may or may not interfere.

t0mdavid-m committed:
only use config?

t0mdavid-m committed:
make cpack filename manually changeable

t0mdavid-m commented:
> Do I understand correctly that you want to specify zip+nsis essentially? And you want to do it because the package_nsis creates some components that you want in the zip on windows?

Yes I want the package to be configured as usual for windows and then package it as zip instead of in an installer. Actually, I am not sure whether all steps are needed for my specific use case / at all (e.g. `include(InstallRequiredSystemLibraries)` discussed in #7528) but this seems like the most stable and consistent implementation for now.

> Maybe I would double check the package_xyz.cmake files then and pull out things that actually are dependent on the OS instead of the actual package type and put it into package_general.cmake.
> 
> I think I gradually tried to do that. Note that there is some code in other cmake sections that depends on packaging (e.g. Mac app bundle creation, which is a bit hindering during development). This may or may not interfere.

This sounds like a good idea. I will gradually start cleaning it up.

t0mdavid-m commented:
I also pushed another commit, which allows for the manual selection of the final package name. The default package name changes dynamically, which can be hindering when used within Github Actions.

if `CPACK_PACKAGE_FILE_NAME` is defined as an environment variable, it is used preferentially over the dynamically generated package name.

jpfeuffer commented on cmake/package_deb.cmake:
```diff
@@ -31,13 +31,17 @@
+  if (OPENMS_64BIT_ARCHITECTURE)
+    set(CPACK_PACKAGE_FILE_NAME "${CPACK_PACKAGE_NAME}-${OPENMS_PACKAGE_VERSION_FULLSTRING}-Debian-Linux-x86_64")
+  else()
+    set(CPACK_PACKAGE_FILE_NAME "${CPACK_PACKAGE_NAME}-${OPENMS_PACKAGE_VERSION_FULLSTRING}-Debian-Linux-x86")
```
Do Debian versions that we support still support x86?
I think arm64 would be more important.
Ah but I see: it was there before. If you have time you can try to support arm here and/or remove it.

jpfeuffer commented on cmake/package_nsis.cmake:
```diff
@@ -34,22 +34,25 @@
-else()
-  MESSAGE(STATUS "Found 8k version of NSIS. Great!")
+# Check if nsis is actually run or if only the configuration is used
+if("${CPACK_GENERATOR}" STREQUAL "NSIS")
```
Yeah, you could avoid that if everything in package_nsis was really just related to the package type, not the OS. But you can check what is needed first.

cbielow commented on CMakeLists.txt:
```diff
@@ -486,9 +486,27 @@ install_export_targets()
+      set(CPACK_GENERATOR "DEB")
+    elseif("${PACKAGE_TYPE}" STREQUAL "nsis")
+      set(CPACK_GENERATOR "NSIS")
+    endif()
```
Does it make sense to add a default case, and throw an error?

jpfeuffer replied:
And in the case you wait for the restructure to detangle package type from OS, maybe warn if someone tries to set "zip" in the usual way via package_type?
"Warning please use it like this: .."

t0mdavid-m replied:
I added default cases to throw errors and added a warning if zip is set as a package type. Thanks a lot for the suggestion!

cbielow reviewed (commented):
As an orthogonal idea (I'm not sure this covers all problems, but its very easy to do):

the NSIS package is a zip file (a self-extracting one, but nevertheless).
It contains all executables plus dependent libs in the ./bin subdirectory.
You can extract that using 7-zip on the commandline.
Same procedure for ./share. 

try e.g. https://github.com/OpenMS/OpenMS/releases/download/Release3.1.0/OpenMS-3.1.0-Win64.exe

timosachsenberg commented on contrib:
anything up with the contrib?

t0mdavid-m replied:
Nothing up with contrib. Just a mistake from my side. Thank you bringing it to my attention!

t0mdavid-m committed:
reset changes in contrib

t0mdavid-m commented:
> the NSIS package is a zip file (a self-extracting one, but nevertheless).
> It contains all executables plus dependent libs in the ./bin subdirectory.
> You can extract that using 7-zip on the commandline.
> Same procedure for ./share.

This is definitely a convenient trick I did not know about. However, I think having an explicit option for zip packaging might be more straightforward from a user perspective. What do you think?

timosachsenberg commented:
I agree. Let’s say we move to wix at some point we probably could not do this trick anymore…

t0mdavid-m committed:
add error messages

t0mdavid-m committed:
change error to warning

t0mdavid-m commented:
I addressed all the "low hanging fruits" for now. I think cleaning up and potentially restructuring the os dependent cmake files is a great idea. However, I think for the streamlit apps it would be best to include the PR in the next release (The windows versions do not work without the dependencies included). 

Thus, I would suggest to open an issue / new PR where I address the restructuring and merge this PR for now.

t0mdavid-m committed:
Merge remote-tracking branch 'OpenMS/develop' into add_zip_option

t0mdavid-m committed:
only support windows

t0mdavid-m commented:
I just had a discussion with @timosachsenberg. Supporting the zip packaging might not make sense for all platforms. Hence, I adjusted the script as follows:

- I removed the `PACK_ZIP` environment variable
- If `PACKAGE_TYPE` is set to `zip`a zip package is built using the `package_nsis.cmake` script
- If `PACKAGE_TYPE` is `zip` and the os is not windows an error is thrown

I still want to do the refactoring as discussed but I would propose setting up a new PR for that such that the new functionality can be included in the `3.2.0` release (see my previous comment).

jpfeuffer commented:
I would argue that it does make sense for all platforms. You just haven't found a use case yet 😅
But yes , if you don't need it right now and everything is backwards compatible I don't see a reason not to merge.

timosachsenberg commented:
haha ok. fair enough... I mean it is a temporary solution right now.

My reasoning (if it makes sense) was that the extra env variable (in addition to a package type) might add some additional complexity. Our primary aim is currently just to have a simple way to package an archived build+dependencies for the web app installer. Having a package type "zip" that creates zipped and relocatable/"portable" builds for additional platforms would be great in the future for that purpose.
The other way of having an extra Env flag + logic to decide when not to package but zip instead (in case of the Windows package script) seemed a bit strange to me - left alone the semantics for the other platforms were not clear.

timosachsenberg reviewed (approved):
fine with me

t0mdavid-m committed:
fix error handling

t0mdavid-m commented:
I made a small change regarding the error handling as the error I added for `PACKAGE_TYPE` turned out to be redundant. I also verified that the portable installation works with `FLASHViewer`.

Thus, if there are no objections from @jpfeuffer, the PR would be ready to merge from my side once the CI is finished.
